### ISSUE TITLE
Storybook integration

### ISSUE THREAD

**Issue by steve8708** [BuilderIO/mitosis#175]

**Comment by Harkunwar** [comment]
I'd love to contribute to the Mitosis project and this seems like something I can look into.

**Comment by steve8708** [comment]
That would be amazing @Harkunwar!

**Comment by steve8708** [comment]
I will go ahead and assign this to you here @Harkunwar - if you find we should change that (other obligations arise, etc) just lmk

**Comment by Harkunwar** [comment]
Thank you @steve8708! Can I add you on discord, I was having some setup issues with the repo. Also wanted to chat about the possible design for this.

**Comment by steve8708** [comment]
Absolutely! I'm `steve8708` on discord and we also have a server with a Mitosis channel [here](https://discord.gg/QPJuae9Ryc)

**Comment by steve8708** [comment]
hey @Harkunwar - have you had a chance to make progress on this one? Mostly just asking to see if you haven't had the time available, I can unassign you and see if someone else is able to pick this up

**Comment by PatrickJS** [comment]
yeah this is needed

**Comment by PatrickJS** [comment]
I can work on a public webpack  mitosis-loader/plugin and webpack playground to show raw, compiled components. We created a sample-page internally to check core mitosis changes that helpped a lot too. We also used jsfiddle on the output to what we think the generator should be for webcomponent so i can incorporate something like that too

**Comment by Harkunwar** [comment]
Hello everyone! Sorry I've been busy lately but I have done some preliminary research on this topic and I still plan to work on this. I'll start creating MRs with my progress soon. I've spent most of my last two months writing esbuild plugins just to familiarize myself with JS compiling and hopefully now I can use that knowledge here.

**Comment by steve8708** [comment]
Sounds great, can't wait @Harkunwar !

**Comment by samijaber** [comment]
@Harkunwar Awesome. Feel free to reach out in our Discord with any questions/issues you may have: this is a big project and there'll likely be complications 😅

**Comment by kylecordes** [comment]
Storybook integration is a great idea, but having used Storybook with various component frameworks, I've noticed that the files needed to test these components via storybook are not quite the same (or in some cases pretty different) from what is needed to see if the component compiles and works directly with its framework, with no Storybook involved.

Therefore it may be wise to consider a different testing approach:

Bunch of mitosis components that exercise every feature of mitosis, and exercise any coding patterns that has revealed a bug in the past.

Compile those to all N frameworks.

Hand code the bare minimum for each of the frameworks to make it use those components - this should not be much.

Run the same Playwright test against the resulting applications - I think this would cover all of the web targets. React Native and others may need separate support.

This would avoid having to step through the Storybook nuances for each target framework.

**Comment by ild0tt0re** [comment]
@samijaber 
Storybook already have a built-in test feature based on playwright
https://storybook.js.org/docs/react/writing-tests/interaction-testing

**Comment by TzviPM** [comment]
@Harkunwar are you actively working on this? I’d love to take this over and contribute. My thoughts are as follows:

The common Story Format (CSF) is based on ES6 modules with a default export containing metadata and named exports for each story.

To my knowledge, Mitosis only supports default exports with a single component per file. This leads me to a few initial ideas on how to approach this:

1. Define a directory-based CSF structure with a mitosis file per story and a metadata file.
2. Create a storybook-specific pipeline for mitosis that supports named exports for stories.
3. Add support for named exports to mitosis, in general.

Regardless of the approach taken for generating the stories themselves, I’m thinking that we can take 2 approaches for storybook configs.

1. The first approach is to simply have the user generate the stories for each framework and link to them from a separate config per framework.
2. The second approach is to create a meta-config format that can be used to generate the storybook config for each framework.

I once saw an interesting example with Nx.dev that had a project set up for each of the following:

1. component library A
2. component library B
3. storybook that consumes the library A
4. storybook for library B
5. e2e component tests that use cypress to test the storybook app for A
6. e2e component tests that use cypress to test the storybook app for B
7. An app consuming both component libraries
8. e2e tests for the app

I’m not sure how well multiple storybook projects in a single repo would work, in general, but I’d probably start with the Nx schematics to see how they work.


Overall, I think the simplest path forward is as follows:

1. Write a tool that takes a folder of mitosis components and metadata and generates a CSF file for each framework
2. Users can create their own config for each framework
3. Create a sample repo with this working for a component library

Once this is working, some areas to pursue include:

1. Exploring more convenient ways to define stories in a single file / enable named exports for mitosis
2. Explore meta-configuration for multiple storybook projects based on mitosis
3. Generate a top-level storybook that includes a way to view all stories + components across frameworks
4. Explore approaches to testing multiple parallel stories for different frameworks from a single test definition. For this, perhaps an approach using something like cucumber might be best, where framework-specific adaptors can be created and tests generated from a single specification.

**Comment by Takouaghouaiel** [comment]
Hello, is there any news on this issue?
